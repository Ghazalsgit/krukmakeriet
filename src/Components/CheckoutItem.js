import React, { useContext } from "react";
import styles from "./CheckoutItem.module.css";
import { AllContext } from "../context/AllContext";

function CheckoutItem({ productData }) {
  /*
  let { productBasket} = useContext(AppContext)
  let { courseBasket} = useContext(AppContext)

  
  if (courseBasket === null) {courseBasket = []}
  if (productBasket === null) {productBasket = []}

  */

  const { productBasket, setProductBasket, courseBasket, setCourseBasket } =
    useContext(AllContext);

  const isProduct = courseBasket.some((product) => {
    if (product.id === productData.id) {
      return true;
    }
    return false;
  });

  function decrementAmount(productData) {
    const courseExist = courseBasket.find((item) => item.id === productData.id);
    const productExist = productBasket.find(
      (item) => item.id === productData.id
    );

    if (isProduct) {
      courseExist.amount > 0
        ? setCourseBasket(
            courseBasket.map((item) =>
              item.id === productData.id
                ? { ...courseExist, amount: courseExist.amount + -1 }
                : item
            )
          )
        : console.log("Nothing to remove");
    } else {
      productExist.amount > 0
        ? setProductBasket(
            productBasket.map((item) =>
              item.id === productData.id
                ? { ...productExist, amount: productExist.amount + -1 }
                : item
            )
          )
        : console.log("Nothing to remove");
    }
  }

  function incrementAmount(productData) {
    const productExist = productBasket.find(
      (item) => item.id === productData.id
    );
    setProductBasket(
      productBasket.map((item) =>
        item.id === productData.id
          ? { ...productExist, amount: productExist.amount + 1 }
          : item
      )
    );

    const courseExist = courseBasket.find((item) => item.id === productData.id);
    setCourseBasket(
      courseBasket.map((item) =>
        item.id === productData.id
          ? { ...courseExist, amount: courseExist.amount + 1 }
          : item
      )
    );
  }

  function deleteItem(productData) {
    //.filter på alla som inte är productData.id
    //uppdatera state med den här listan
    setCourseBasket(courseBasket.filter((item) => item.id !== productData.id));
    setProductBasket(
      productBasket.filter((item) => item.id !== productData.id)
    );
  }

  return (
    <article className={styles.productItemWrapper} data-testid="checkoutItem">
      <div className={styles.imgWrapper}>
        <img className={styles.itemImage} src={productData.img} alt="" />
      </div>
      <div className={styles.infoWrapper}>
        <h3 className={styles.name}>{productData.name}</h3>
        <p className={styles.info}>{productData.info}</p>
        <p className={styles.price}>{productData.price}</p>
        <div className={styles.editQuantityWrapper}>
          <button
            data-testid="decrementBtn"
            onClick={() => decrementAmount(productData)}
            className={styles.quantityBtn}
          >
            -
          </button>
          <p data-testid="counterText">{productData.amount}</p>
          <button
            data-testid="incrementBtn"
            onClick={() => incrementAmount(productData)}
            className={styles.quantityBtn}
          >
            +
          </button>
        </div>
      </div>
      <button
        onClick={() => deleteItem(productData)}
        className={styles.removeBtn}
      >
        Delete
      </button>
    </article>
  );
}

export default CheckoutItem;
