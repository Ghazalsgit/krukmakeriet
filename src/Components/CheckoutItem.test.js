import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import CheckoutItem from "./CheckoutItem";
import { AllContextProvider } from "../context/AllContext";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";

const MockCheckoutItem = () => {
  return (
    <BrowserRouter>
      <AllContextProvider>
        <CheckoutItem productData={product} />
      </AllContextProvider>
    </BrowserRouter>
  );
};
const product = { name: "skÃ¥l", img: "https" };

describe("CheckoutItem", () => {
  test("check if props exist as should", async () => {
    render(<MockCheckoutItem />);
    const routeForCheckoutItem = screen.getByTestId("checkoutItem");
    expect(routeForCheckoutItem).toBeInTheDocument();
  });
  test("create a snapshot to test the rendering of element", async () => {
    const domTree = renderer
      .create(
        <BrowserRouter>
          <AllContextProvider>
            <CheckoutItem productData={product} />
          </AllContextProvider>
        </BrowserRouter>
      )
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });

  test("check if onclick on increment button the amount changes", async () => {
    render(<MockCheckoutItem />);
    const incrementBtn = screen.getByTestId("incrementBtn");
    const counterText = screen.getByTestId("counterText");
    expect(counterText).toHaveTextContent("");
    userEvent.click(incrementBtn);
    expect(counterText).toHaveTextContent("1");
  });

  test("check if onclick on decrement button the amount changes", async () => {
    render(<MockCheckoutItem />);
    const decremetnBtn = screen.getByTestId("decrementBtn");
    const counterText = screen.getByTestId("counterText");
    expect(counterText).toHaveTextContent("1");
    userEvent.click(decremetnBtn);
    expect(counterText).toHaveTextContent("");
  });
});
