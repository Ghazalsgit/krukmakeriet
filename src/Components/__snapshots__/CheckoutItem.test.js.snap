// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutItem create a snapshot to test the rendering of element 1`] = `
<article
  className="productItemWrapper"
  data-testid="checkoutItem"
>
  <div
    className="imgWrapper"
  >
    <img
      alt=""
      className="itemImage"
      src="https"
    />
  </div>
  <div
    className="infoWrapper"
  >
    <h3
      className="name"
    >
      sk√•l
    </h3>
    <p
      className="info"
    />
    <p
      className="price"
    />
    <div
      className="editQuantityWrapper"
    >
      <button
        className="quantityBtn"
        data-testid="decrementBtn"
        onClick={[Function]}
      >
        -
      </button>
      <p
        data-testid="counterText"
      />
      <button
        className="quantityBtn"
        data-testid="incrementBtn"
        onClick={[Function]}
      >
        +
      </button>
    </div>
  </div>
  <button
    className="removeBtn"
    onClick={[Function]}
  >
    Delete
  </button>
</article>
`;
